### ISSUE TITLE
Error: "There is no CryptoRepoFile for this cryptoRepoFileId: ..."

### ISSUE THREAD

**Issue by nlmarco** [subshare/subshare#57]
Encountered an out-of-sync problem a while ago (and a friend of mine had a similar (or the same?) problem):

```
2017-06-25 22:29:45,708 [RepoSyncDaemonThread_0_5] ERROR c.c.c.c.r.s.RepoSyncDaemonImpl$WrapperRunnable - run: java.lang.IllegalArgumentException: There is no CryptoRepoFile for this cryptoRepoFileId: ZArm_JSQijGUxmyNxrL0hA
java.lang.IllegalArgumentException: There is no CryptoRepoFile for this cryptoRepoFileId: ZArm_JSQijGUxmyNxrL0hA
        at org.subshare.local.persistence.CryptoRepoFileDao.getCryptoRepoFileOrFail(CryptoRepoFileDao.java:28) ~[org.subshare.local-0.10.0.jar:na]
        at org.subshare.local.CryptreeImpl.putCryptoRepoFileDto(CryptreeImpl.java:1328) ~[org.subshare.local-0.10.0.jar:na]
        at org.subshare.local.CryptreeImpl.putCryptoChangeSetDto(CryptreeImpl.java:375) ~[org.subshare.local-0.10.0.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.syncCryptoKeysFromRemoteRepo(CryptreeRestRepoTransportImpl.java:246) ~[org.subshare.rest.client-0.10.0.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.getChangeSetDto(CryptreeRestRepoTransportImpl.java:168) ~[org.subshare.rest.client-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.sync(RepoToRepoSync.java:223) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.syncDown(RepoToRepoSync.java:174) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.sync(RepoToRepoSync.java:137) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at org.subshare.core.repo.sync.SsRepoToRepoSync.sync(SsRepoToRepoSync.java:65) ~[org.subshare.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoSyncRunner.run(RepoSyncRunner.java:51) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoSyncDaemonImpl$WrapperRunnable.run(RepoSyncDaemonImpl.java:123) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_111]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_111]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_111]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_111]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_111]
```

I assume that this is caused by the server storing a higher "last sync revision" for the client-repo as the client really has. Maybe the client's DB transaction failed (though this should be first locally committed and then sent to the client). I have no idea, what this problem actually caused, but I hope that #53 prevents it for the future.

For the broken "out-of-sync" repo, we need a repair possibility.

**Comment by nlmarco** [comment]
I deployed the new snapshot to the productive server and to my (affected) notebook, but even after resetting the last-sync-revision (via the newly introduced debug-/admin-feature) and re-downloading all crypto-meta-data, the error still happens :-(

I analysed the above stack-trace and the corresponding sources again and assume that the problem is actually different: It seems that the parent-CryptoRepoFileDto and the currently put CryptoRepoFileDto are in the same CryptoChangeSet and the parent was not yet written into the DB (the child comes first).

I just added code to sort CryptoChangeSet.cryptoRepoFileDtos before iterating and writing them to the DB. I hope that this solves the problem.
